File: app/src/main/java/com/botijasoftware/utils/materials/Material.java
package com.botijasoftware.utils.materials;

import java.util.ArrayList;
import com.botijasoftware.utils.ColorRGB;
import com.botijasoftware.utils.ColorRGBA;
import com.botijasoftware.utils.ResourceManager;

import android.opengl.GLES10;

public class Material {

	public ArrayList<MaterialPass> mPass;
	public ColorRGBA mDiffuseColor;
	public ColorRGB mAmbientColor;
	public ColorRGB mEmissionColor;
	public float mShininess;
	public ColorRGB mSpecularColor;
	private boolean loaded = false; 
	private float color[] = new float[4];
	
	public static final int MAX_GLOBAL_PARAMETERS = 16;
	public static final int MAX_LOCAL_PARAMETERS = 4;
	public static float[] mGlobalArgs = new float[MAX_GLOBAL_PARAMETERS];
	public float[] mLocalArgs = new float[MAX_LOCAL_PARAMETERS];
	public static float time;
	
	
	public final static int MAT_VISIBLE = 0x01;
	public final static int MAT_NOSHADOWS = 0x02;
	public final static int MAT_NODEPTHTEST = 0x04;
	public final static int MAT_TRANSLUCENT = 0x08;
	public final static int MAT_ALPHATEST = 0x10;
	public final static int MAT_MASKDEPTH = 0x20;
	public final static int MAT_NOBLEND = 0x40;
	public final static int MAT_NOLIGHTS = 0x80;
	public final static int MAT_FLAT = 0x100;
	public final static int MAT_BACKSIDED = 0x200;
	public final static int MAT_TWOSIDED = 0x400;
	public final static int MAT_POLYGONOFFSET = 0x800;
	
	
	public final static int SORT_BACKGROUND = 0;
	public final static int SORT_SOLID = 1;
	public final static int SORT_TRANSPARENT = 2;
	public final static int SORT_SCREEN = 3;
	public final static int SORT_POSTPROCESS = 4;

	
	public int sort = SORT_SOLID;
	public int mask = 0;
	public float alphatest = 1.0f;
	public float polygonoffset = 0.0f;
	
	public void setMask(int maskbits) {
		mask = maskbits;
	}
	
	public void addMask(int maskbits) {
		mask |= maskbits;
	}
	
	public int getMask() {
		return mask;
	}
	
	public boolean checkMask(int maskbits) {
		return ((mask & maskbits) != 0);
	}
	
	
	public Material() {
		mPass = new ArrayList<MaterialPass>();
		mDiffuseColor = new ColorRGBA(1,1,1,1);
		mAmbientColor = new ColorRGB(1,1,1);
		mShininess = 128f;
		mSpecularColor = new ColorRGB(1,1,1);
		mEmissionColor = new ColorRGB(1,1,1);
	}
	
	public void LoadContent(ResourceManager resources) {
		for (int i = 0; i< mPass.size();i++)
			mPass.get(i).LoadContent(resources);
		
		loaded = true;
	}
	
	public void setDiffuseColor(float r, float g, float b) {
		mDiffuseColor.R = r;
		mDiffuseColor.G = g;
		mDiffuseColor.B = b;
	}
	
	public void setAmbientColor(float r, float g, float b) {
		mAmbientColor.R = r;
		mAmbientColor.G = g;
		mAmbientColor.B = b;
	}
	
	public void setShininess(float shininess) {
		mShininess = shininess;
	}

	public void setSpecularColor(float r, float g, float b) {
		mSpecularColor.R = r;
		mSpecularColor.G = g;
		mSpecularColor.B = b;
	}
	
	public void setEmissionColor(float r, float g, float b) {
		mEmissionColor.R = r;
		mEmissionColor.G = g;
		mEmissionColor.B = b;
	}
	
	
	public void set() {
		
		if ((mask & MAT_FLAT) != 0 ) GLES10.glShadeModel(GLES10.GL_FLAT);
		if ((mask & MAT_TWOSIDED) != 0 ) GLES10.glDisable(GLES10.GL_CULL_FACE);
		else if ((mask & MAT_BACKSIDED) != 0 ) GLES10.glCullFace(GLES10.GL_CCW);
		if ((mask & MAT_NODEPTHTEST) != 0 ) GLES10.glDisable(GLES10.GL_DEPTH_TEST);
		if ((mask & MAT_MASKDEPTH) != 0 ) GLES10.glDepthMask(true);
		if ((mask & MAT_NOBLEND) != 0 ) GLES10.glDisable(GLES10.GL_BLEND);
		if ((mask & MAT_NOLIGHTS) != 0 ) GLES10.glDisable(GLES10.GL_LIGHTING);
		if ((mask & MAT_ALPHATEST) != 0 ) {
			GLES10.glEnable(GLES10.GL_ALPHA_TEST);
			GLES10.glAlphaFunc(GLES10.GL_GREATER, alphatest);
		}
		
		if ((mask & MAT_POLYGONOFFSET) != 0 ) {
			GLES10.glEnable(GLES10.GL_POLYGON_OFFSET_FILL);
			GLES10.glPolygonOffset(0.0f, polygonoffset);
		}
		
		GLES10.glDepthFunc(GLES10.GL_LEQUAL);
		
		GLES10.glColor4f(mDiffuseColor.R, mDiffuseColor.G, mDiffuseColor.B, mDiffuseColor.A);
		/*color[0] = mDiffuseColor.R;
		color[1] = mDiffuseColor.G;
		color[2] = mDiffuseColor.B;
		color[3] = 1.0f;
		GLES10.glMaterialfv(GLES10.GL_FRONT_AND_BACK, GLES10.GL_DIFFUSE, color,0);
		color[0] = mAmbientColor.R;
		color[1] = mAmbientColor.G;
		color[2] = mAmbientColor.B;		
		GLES10.glMaterialfv(GLES10.GL_FRONT_AND_BACK, GLES10.GL_AMBIENT, color,0);
		color[0] = mSpecularColor.R;
		color[1] = mSpecularColor.G;
		color[2] = mSpecularColor.B;		
		GLES10.glMaterialfv(GLES10.GL_FRONT_AND_BACK, GLES10.GL_SPECULAR, color,0);
		color[0] = mEmissionColor.R;
		color[1] = mEmissionColor.G;
		color[2] = mEmissionColor.B;		
		GLES10.glMaterialfv(GLES10.GL_FRONT_AND_BACK, GLES10.GL_EMISSION, color,0);
		GLES10.glMaterialf(GLES10.GL_FRONT_AND_BACK, GLES10.GL_SHININESS, mShininess);
		GLES10.glEnable(GLES10.GL_COLOR_MATERIAL);*/
		
	}
	
	public void unSet() {
		
		if ((mask & MAT_FLAT) != 0 ) GLES10.glShadeModel(GLES10.GL_SMOOTH);
		if ((mask & MAT_TWOSIDED) != 0 )GLES10.glEnable(GLES10.GL_CULL_FACE);
		else if ((mask & MAT_BACKSIDED) != 0 ) GLES10.glCullFace(GLES10.GL_CW);
		if ((mask & MAT_NODEPTHTEST) != 0 ) GLES10.glEnable(GLES10.GL_DEPTH_TEST);
		if ((mask & MAT_MASKDEPTH) != 0 ) GLES10.glDepthMask(false);
		if ((mask & MAT_NOBLEND) != 0 ) GLES10.glEnable(GLES10.GL_BLEND);
		if ((mask & MAT_NOLIGHTS) != 0 ) GLES10.glEnable(GLES10.GL_LIGHTING);
		if ((mask & MAT_ALPHATEST) != 0 ) GLES10.glDisable(GLES10.GL_ALPHA_TEST);
		if ((mask & MAT_POLYGONOFFSET) != 0 ) GLES10.glDisable(GLES10.GL_POLYGON_OFFSET_FILL);
		
		GLES10.glColor4f(1, 1, 1, 1);
		//GLES10.glMaterialf(GLES10.GL_FRONT_AND_BACK, GLES10.GL_SHININESS, 1);
		//GLES10.glDisable(GLES10.GL_COLOR_MATERIAL);
		

	}

	public void addPass(MaterialPass pass) {
		mPass.add(pass);
	}
	
	public int passCount() {
		return mPass.size();
	}
	
	public float getTime() {
		return Material.time;
	}
	
	public float getGlobalArg(int index) {
		if (index >0 && index < MAX_GLOBAL_PARAMETERS) {
			return Material.mGlobalArgs[index];
		}
		return 0.0f;
	}
	
	public float getLocalArg(int index) {
		if (index >0 && index < MAX_LOCAL_PARAMETERS) {
			return mLocalArgs[index];
		}
		return 0.0f;
	}
	
	public void setGlobalArg(int index, float val) {
		if (index >0 && index < MAX_GLOBAL_PARAMETERS) {
			Material.mGlobalArgs[index] = val;
		}
	}
	
	public void setLocalArg(int index, float val) {
		if (index >0 && index < MAX_LOCAL_PARAMETERS) {
			mLocalArgs[index] = val;
		}
	}

	public boolean isLoaded() {
		return loaded;
	}
}

File: app/src/main/java/com/botijasoftware/utils/materials/Material.kt
package com.botijasoftware.utils.materials

import java.util.ArrayList
import com.botijasoftware.utils.ColorRGB
import com.botijasoftware.utils.ColorRGBA
import com.botijasoftware.utils.ResourceManager

import android.opengl.GLES10

class Material {

    var mPass: ArrayList<MaterialPass>
    var mDiffuseColor: ColorRGBA
    var mAmbientColor: ColorRGB
    var mEmissionColor: ColorRGB
    var mShininess: Float = 0.toFloat()
    var mSpecularColor: ColorRGB
    var isLoaded = false
        private set
    private val color = FloatArray(4)
    var mLocalArgs = FloatArray(MAX_LOCAL_PARAMETERS)


    var sort = SORT_SOLID
    var mask = 0
    var alphatest = 1.0f
    var polygonoffset = 0.0f

    fun addMask(maskbits: Int) {
        mask = mask or maskbits
    }

    fun checkMask(maskbits: Int): Boolean {
        return mask and maskbits != 0
    }

    init {
        mPass = ArrayList<MaterialPass>()
        mDiffuseColor = ColorRGBA(1f, 1f, 1f, 1f)
        mAmbientColor = ColorRGB(1f, 1f, 1f)
        mShininess = 128f
        mSpecularColor = ColorRGB(1f, 1f, 1f)
        mEmissionColor = ColorRGB(1f, 1f, 1f)
    }

    fun LoadContent(resources: ResourceManager) {
        for (i in mPass.indices)
            mPass[i].LoadContent(resources)

        isLoaded = true
    }

    fun setDiffuseColor(r: Float, g: Float, b: Float) {
        mDiffuseColor.R = r
        mDiffuseColor.G = g
        mDiffuseColor.B = b
    }

    fun setAmbientColor(r: Float, g: Float, b: Float) {
        mAmbientColor.R = r
        mAmbientColor.G = g
        mAmbientColor.B = b
    }

    fun setShininess(shininess: Float) {
        mShininess = shininess
    }

    fun setSpecularColor(r: Float, g: Float, b: Float) {
        mSpecularColor.R = r
        mSpecularColor.G = g
        mSpecularColor.B = b
    }

    fun setEmissionColor(r: Float, g: Float, b: Float) {
        mEmissionColor.R = r
        mEmissionColor.G = g
        mEmissionColor.B = b
    }


    fun set() {

        if (mask and MAT_FLAT != 0) GLES10.glShadeModel(GLES10.GL_FLAT)
        if (mask and MAT_TWOSIDED != 0)
            GLES10.glDisable(GLES10.GL_CULL_FACE)
        else if (mask and MAT_BACKSIDED != 0) GLES10.glCullFace(GLES10.GL_CCW)
        if (mask and MAT_NODEPTHTEST != 0) GLES10.glDisable(GLES10.GL_DEPTH_TEST)
        if (mask and MAT_MASKDEPTH != 0) GLES10.glDepthMask(true)
        if (mask and MAT_NOBLEND != 0) GLES10.glDisable(GLES10.GL_BLEND)
        if (mask and MAT_NOLIGHTS != 0) GLES10.glDisable(GLES10.GL_LIGHTING)
        if (mask and MAT_ALPHATEST != 0) {
            GLES10.glEnable(GLES10.GL_ALPHA_TEST)
            GLES10.glAlphaFunc(GLES10.GL_GREATER, alphatest)
        }

        if (mask and MAT_POLYGONOFFSET != 0) {
            GLES10.glEnable(GLES10.GL_POLYGON_OFFSET_FILL)
            GLES10.glPolygonOffset(0.0f, polygonoffset)
        }

        GLES10.glDepthFunc(GLES10.GL_LEQUAL)

        GLES10.glColor4f(mDiffuseColor.R, mDiffuseColor.G, mDiffuseColor.B, mDiffuseColor.A)
        /*color[0] = mDiffuseColor.R;
		color[1] = mDiffuseColor.G;
		color[2] = mDiffuseColor.B;
		color[3] = 1.0f;
		GLES10.glMaterialfv(GLES10.GL_FRONT_AND_BACK, GLES10.GL_DIFFUSE, color,0);
		color[0] = mAmbientColor.R;
		color[1] = mAmbientColor.G;
		color[2] = mAmbientColor.B;
		GLES10.glMaterialfv(GLES10.GL_FRONT_AND_BACK, GLES10.GL_AMBIENT, color,0);
		color[0] = mSpecularColor.R;
		color[1] = mSpecularColor.G;
		color[2] = mSpecularColor.B;
		GLES10.glMaterialfv(GLES10.GL_FRONT_AND_BACK, GLES10.GL_SPECULAR, color,0);
		color[0] = mEmissionColor.R;
		color[1] = mEmissionColor.G;
		color[2] = mEmissionColor.B;
		GLES10.glMaterialfv(GLES10.GL_FRONT_AND_BACK, GLES10.GL_EMISSION, color,0);
		GLES10.glMaterialf(GLES10.GL_FRONT_AND_BACK, GLES10.GL_SHININESS, mShininess);
		GLES10.glEnable(GLES10.GL_COLOR_MATERIAL);*/

    }

    fun unSet() {

        if (mask and MAT_FLAT != 0) GLES10.glShadeModel(GLES10.GL_SMOOTH)
        if (mask and MAT_TWOSIDED != 0)
            GLES10.glEnable(GLES10.GL_CULL_FACE)
        else if (mask and MAT_BACKSIDED != 0) GLES10.glCullFace(GLES10.GL_CW)
        if (mask and MAT_NODEPTHTEST != 0) GLES10.glEnable(GLES10.GL_DEPTH_TEST)
        if (mask and MAT_MASKDEPTH != 0) GLES10.glDepthMask(false)
        if (mask and MAT_NOBLEND != 0) GLES10.glEnable(GLES10.GL_BLEND)
        if (mask and MAT_NOLIGHTS != 0) GLES10.glEnable(GLES10.GL_LIGHTING)
        if (mask and MAT_ALPHATEST != 0) GLES10.glDisable(GLES10.GL_ALPHA_TEST)
        if (mask and MAT_POLYGONOFFSET != 0) GLES10.glDisable(GLES10.GL_POLYGON_OFFSET_FILL)

        GLES10.glColor4f(1f, 1f, 1f, 1f)
        //GLES10.glMaterialf(GLES10.GL_FRONT_AND_BACK, GLES10.GL_SHININESS, 1);
        //GLES10.glDisable(GLES10.GL_COLOR_MATERIAL);


    }

    fun addPass(pass: MaterialPass) {
        mPass.add(pass)
    }

    fun passCount(): Int {
        return mPass.size
    }

    fun getTime(): Float {
        return Material.time
    }

    fun getGlobalArg(index: Int): Float {
        if (index > 0 && index < MAX_GLOBAL_PARAMETERS) {
            return Material.mGlobalArgs[index]
        }
        return 0.0f
    }

    fun getLocalArg(index: Int): Float {
        if (index > 0 && index < MAX_LOCAL_PARAMETERS) {
            return mLocalArgs[index]
        }
        return 0.0f
    }

    fun setGlobalArg(index: Int, `val`: Float) {
        if (index > 0 && index < MAX_GLOBAL_PARAMETERS) {
            Material.mGlobalArgs[index] = `val`
        }
    }

    fun setLocalArg(index: Int, `val`: Float) {
        if (index > 0 && index < MAX_LOCAL_PARAMETERS) {
            mLocalArgs[index] = `val`
        }
    }

    companion object {

        val MAX_GLOBAL_PARAMETERS = 16
        val MAX_LOCAL_PARAMETERS = 4
        var mGlobalArgs = FloatArray(MAX_GLOBAL_PARAMETERS)
        var time: Float = 0.toFloat()


        val MAT_VISIBLE = 0x01
        val MAT_NOSHADOWS = 0x02
        val MAT_NODEPTHTEST = 0x04
        val MAT_TRANSLUCENT = 0x08
        val MAT_ALPHATEST = 0x10
        val MAT_MASKDEPTH = 0x20
        val MAT_NOBLEND = 0x40
        val MAT_NOLIGHTS = 0x80
        val MAT_FLAT = 0x100
        val MAT_BACKSIDED = 0x200
        val MAT_TWOSIDED = 0x400
        val MAT_POLYGONOFFSET = 0x800


        val SORT_BACKGROUND = 0
        val SORT_SOLID = 1
        val SORT_TRANSPARENT = 2
        val SORT_SCREEN = 3
        val SORT_POSTPROCESS = 4
    }
}
